ramp-engine/ramp_engine/aws/api.py

```diff
@@ -795,9 +795,14 @@ def _is_ready(config, instance_id):
 
 def _training_finished(config, instance_id, submission_name):
     """
-    Return True if a submission has finished training
+    Return True if a submission has finished training (if the screen no longer
+    exists on the ec2 instance and if the bagged_scores.csv file was saved)
     """
-    return not _has_screen(config, instance_id, submission_name)
+    has_screen = _has_screen(config, instance_id, submission_name)
+    # this can only work if the training was successful
+    has_score_file = _has_error_or_score_file(
+        config, instance_id, submission_name)
+    return not has_screen and has_score_file
```

Comment:
doesn't this risk of deadlocking the workers if a screen failed but the result were not saved? which mechanism ensure that you wi exit the training loop in this case?

Reply:
You are right. Do you have a suggestion for additional check to avoid that?

Reply:
you need to have three states. One `running` (screen alive), one is `finished` (no screen and score file exist) and another one is `broken` (no screen and no score). and stop when you are not `running` anymore.

Reply:
yes. I understand the three states + the `finished` can end on the score file or error file. But the `broken` state is problematic because I don't know how to check if we are not running anymore if we are `broken`. That's why checking if the screen exists was introduced. 
Do you have an idea how to do that?

Otherwise, I want to introduce timeout to AWS instances, that would be partial solution: in those few cases when for some unexplained reason the instance has no screen and it didn't manage to save a file (ie is`broken`) in the worse case scenario it would run for the set `timeout` length of time doing nothing.

Reply:
I would say if you are in a broken state, just restart consider it a a checking error no?
That way you will launch back the submission as you are in a state where you don't know what to do.
This is the classical approach taken in multiprocessing computations.

Reply:
Yes, good idea. We can do that. But still.. do you have idea how to ensure that we are `broken`?
(`no screen` and `no score` previously meant still being in the process of saving the score so it would not be a good idea to only take those two flags into account to restart the whole submission).

Reply:
can't you compute the score in the same screen?
Else, you can use a patience parameter to allow for some time.